### PR TITLE
Issue #3260885 by tBKoT: New conent list block for multiple content types

### PR DESCRIPTION
**Pull request by dmytriikaiun**
## Problem
As default, the custom content list block is allowed to select only one content type (topic, event, group, etc.) but CM+ should be able to select a multiple of them. However, if we just make content type selector multiple it could provide a lot of problems because every content type has its own query conditions.

## Solution
Create a new block that will provide the possibility to select multiple content types with fewer filter/sorting processes.

## Issue tracker
- https://getopensocial.atlassian.net/browse/YANG-6946
- https://www.drupal.org/project/social/issues/3260885

## How to test
- [x] Using the latest version of Open Social with the `social_content_block_landing_page` module enabled
- [x] As a CM+ create a Landing Page
- [x] Add the `Custom multiple content list` section
- [ ] Set up content type and filter/sorting
- [ ] You should be able to see content list.

#### Block configuration
зображення

#### Block display
зображення

## Release notes
Now CM+ users could show a list of multiple content types.

## Change Record
To make it possible to use other content types you should provide a new `@MultipleContentBlock` plugin. 
The `entity_type` and `label` are required for correct work. 
To use only a specific bundle we should add a `bundle` in the annotation, in another case, all entities of the provided `entity_type` will be selected.

## Translations
N/A

### CODE REVIEW

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2723-0eyzxu5r0d2sgpdeaf4xdqxybnnsuvsx.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61f3f4d1046a4b8d294c43e4

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2723-0eyzxu5r0d2sgpdeaf4xdqxybnnsuvsx.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61f3f4d1046a4b8d294c43e4

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2723-0eyzxu5r0d2sgpdeaf4xdqxybnnsuvsx.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61f3f4d1046a4b8d294c43e4

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2723-0eyzxu5r0d2sgpdeaf4xdqxybnnsuvsx.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61f3f4d1046a4b8d294c43e4

File: modules/social_features/social_content_block/src/Plugin/Block/MultipleContentBlock.php [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+    if (!empty($this->getSelectedContentTags())) {
+      $content_tag_table = "{$type}__social_tagging";
+      if ($this->database->schema()->tableExists($content_tag_table)) {
+        $tags = array_column($this->getSelectedContentTags(), 'target_id');
+        $query->innerJoin($content_tag_table, 'st', "st.entity_id = base_table.$id_key");
+        $query->condition('st.social_tagging_target_id', $tags, 'IN');
+      }
+    }
+
+    // Sor by provided sorting field.
```
Comment (Ressinel) [inline modules/social_features/social_content_block/src/Plugin/Block/MultipleContentBlock.php:327]: `Sor` -> `Sort`

**Review by Ressinel — CHANGES_REQUESTED** [review]

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2723-0eyzxu5r0d2sgpdeaf4xdqxybnnsuvsx.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61f3f4d1046a4b8d294c43e4

File: modules/social_features/social_content_block/src/Plugin/Block/MultipleContentBlock.php [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+      '#title' => $this->t('Sorting'),
+    ];
+
+    $form['sorting']['sorting'] = [
+      '#type' => 'select',
+      '#title' => $this->t('Sorting'),
+      '#required' => TRUE,
+      '#options' => [
+        'changed' => $this->t('Last updated'),
+        'created' => $this->t('Most resent'),
```
Comment (rollki) [inline modules/social_features/social_content_block/src/Plugin/Block/MultipleContentBlock.php:162]: Most recent :)

File: modules/social_features/social_content_block/src/Plugin/Block/MultipleContentBlock.php [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+      }
+      else {
+        $types[$definition['entity_type']] = NULL;
+      }
+    }
+
+    // Select all entities from multiple content types.
+    foreach ($types as $type => $bundles) {
+      $entities_data = array_merge($entities_data, $this->getEntitiesByDefinitions((string) $type, $bundles));
+      ${$type . '_storage'} = $this->entityTypeManager->getStorage($type);
```
Comment (rollki) [inline modules/social_features/social_content_block/src/Plugin/Block/MultipleContentBlock.php:257]: :)

File: modules/social_features/social_content_block/src/Plugin/Block/MultipleContentBlock.php [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+    // Sort the list of provided entities.
+    arsort($entities_data);
+    $iteration = 0;
+
+    foreach (array_keys($entities_data) as $key) {
+      if ($iteration >= $this->getAmountItems()) {
+        break;
+      }
+      $keys = explode('__', $key);
+      $entity = ${$keys[0] . '_storage'}->load($keys[1]);
```
Comment (rollki) [inline modules/social_features/social_content_block/src/Plugin/Block/MultipleContentBlock.php:269]: :)

File: modules/social_features/social_content_block/src/Plugin/Block/MultipleContentBlock.php [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+   *
+   * @param string $type
+   *   The content type.
+   * @param array|null $bundles
+   *   List of bundes.
+   *
+   * @return array
+   *   List of entities for render.
+   */
+  protected function getEntitiesByDefinitions(string $type, array $bundles = NULL): array {
```
Comment (rollki) [inline modules/social_features/social_content_block/src/Plugin/Block/MultipleContentBlock.php:295]:
```suggestion
  protected function getEntitiesByDefinitions(string $type, ?array $bundles): array {
```

**Review by rollki — CHANGES_REQUESTED** [review]
In general very good job, only few very small comments from me.

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2723-0eyzxu5r0d2sgpdeaf4xdqxybnnsuvsx.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61f3f4d1046a4b8d294c43e4

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2723-0eyzxu5r0d2sgpdeaf4xdqxybnnsuvsx.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61f3f4d1046a4b8d294c43e4

**Review by Ressinel — COMMENTED** [review]
Hi, @dmytriikaiun Code looks good to me! Nice work!

But I cannot add any content type except **Topic** [here](https://pr2723-0eyzxu5r0d2sgpdeaf4xdqxybnnsuvsx.tugboat.qa/node/24). There is a lot of different content on installation. Could you check this [Landing page](https://pr2723-0eyzxu5r0d2sgpdeaf4xdqxybnnsuvsx.tugboat.qa/node/24) and let me know if I do something wrong?

**Comment by dmytriikaiun** [comment]
@Ressinel 
Yes, as default only Topics can be added all other content types are in submodules. I used the same behavior as for the `Custom content list block`

**Review by rollki — APPROVED** [review]
LGTM.

**Review by Ressinel — APPROVED** [review]

**Comment by rollki** [comment]
Hi @dmytriikaiun 
Could you please rebase commits?  And then we will merge it.

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2723-0eyzxu5r0d2sgpdeaf4xdqxybnnsuvsx.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61f3f4d1046a4b8d294c43e4

**Pull request merged**